hip_data_tools/hipages/version_tracking.py

```diff
@@ -0,0 +1,361 @@
+
+
+def _instantiate_repo(repo_path):
+    """
+    Instaniates a git repo job at the the given path
+    Args:
+        repo_path: absolute or relative path the github repo
+
+    Returns: github repo object
+
+    """
+    return git.Repo(repo_path)
+
+
+def find_all_modules_which_require_version_tracking(file_list):
+    """
+    Iterates through a list of python files, and returns the names of classes,
+    methods and functions and their file locations, which require a version
+    to be found
+    Args:
+        file_list: list of paths to python files
+
+    Returns:
+            defitions_with_tags : list of class names which require version
+            tracking
+            files_with_tag: list of files in which the version tracked files
+            are found
+
+    """
+    files_with_tag = []
+    defitions_with_tags = []
+
+    for path in file_list:
+        defs_with_tags_in_file = \
+            find_any_relevant_decorations_in_file(defitions_with_tags)
+
+        defitions_with_tags = defitions_with_tags + defs_with_tags_in_file
+
+        if defs_with_tags_in_file:
+            files_with_tag.extend(
+                [path for _ in defs_with_tags_in_file]
+            )
+
+    return defitions_with_tags, files_with_tag
+
+
+def find_any_relevant_decorations_in_file(path):
+    """
+    Find any relevant decorations in the provided in tehe
+    Args:
+        path (string): Path to the file being scanned
+
+    Returns (list(string)): List of decorated definitions found in the file
+
+    """
+    classes_in_file_with_decorator = []
+    definitions_with_tags = []
+    for declaration, decorating_string in DEFINITION_MAPPING.items():
+        classes_in_file_with_decorator.extend(
+            check_for_decorated_declaration_in_file(
+                path,
+                decorating_string,
+                declaration))
+    definitions_with_tags.extend(classes_in_file_with_decorator)
+
+    return definitions_with_tags
+
+
+def check_for_decorated_declaration_in_file(path,
+                                            decorating_string,
+                                            declaration):
+    """
+    Finds any classes in a file with the relevant decorator string
+    Args:
+        path: absolute or relative path to the python file
+        decorating_string (string) : string of decorator used to identify the
+             methods, classes and functions which need tracked
+        declaration (string): Declaration to look for
+    Returns: list of classes within the file which require version tracking
+
+    """
+
+    with open(path, 'r') as file:
+        lines = file.readlines()
+        tagged_delcarations = _find_declarations_in_lines(declaration,
+                                                          decorating_string,
+                                                          lines)
+
+    return tagged_delcarations
+
+
+def _find_declarations_in_lines(declaration, decorating_string, lines):
```

Comment:
yeah, I agree this is a bit complex. A few things that might help:
- you use next_line as a proxy to look ahead and read the next line, instead of the if condition, rather use the index and look ahead in the array by lines[current_line_number + 1] that makes it more explicit and easy to understand.
- You can extract out functions that just match / identify the occurances of your keywords, then use the list of indices returned by those occurrences, to read the next lines. then you could simply use that list comprehensions, make it more functional.

Reply:
ah yeah, I agree this is a little complex - i had trouble trying to refactor correctly though.

The problem i was trying to counter is that you can have multiple decorators, so it's just looking for the next line with right keyword in it. So that boolean stays high until it find the correct line.